### PR TITLE
Add OptIn mode for source generator

### PR DESCRIPTION
**Pull request by manodasanW**
This change adds a couple new modes for the source generator.  The source generator can now either run in `Auto` mode which eventually would be a build task or in `OptIn` mode.  By default, in the default mode of true, in WinUI / UWP scenarios, it would run in `Auto` mode.  Whereas in the other scenarios, it would run in `OptIn` mode.  This allows to address the feedback where it was source generating vtables in scenarios where it is probably not needed.  To enable the `OptIn` mode, the `GeneratedWinRTExposedType` and the `GeneratedWinRTExposedExternalType` attributes are added which allows to generate the vtables based on the use of the attributes.  These attributes are added only for the .NET 8 and later versions given the upcoming .NET 6 EOL.

Contributes to #1814

### CODE REVIEW

File: src/WinRT.Runtime/ComWrappersSupport.net5.cs [hunk 1]
```diff
@@ -178,6 +178,10 @@ internal static IntPtr CreateCCWForObjectForABI(object obj, Guid iid)
                 Marshal.ThrowExceptionForHR(Marshal.QueryInterface(ccw, ref iid, out var iidCcw));
                 return iidCcw;
             }
+            catch (InvalidCastException ex)
```
Comment (Sergio0694) [inline src/WinRT.Runtime/ComWrappersSupport.net5.cs:181]: Could we perhaps do something like this instead?

```csharp
IntPtr ccw = ComWrappers.GetOrCreateComInterfaceForObject(obj, CreateComInterfaceFlags.TrackerSupport);

// Invoke 'IUnknown::QueryInterface' on the CCW
int hr = Marshal.QueryInterface(ccw, ref iid, out IntPtr ppv);

// Always release the original CCW after the call
_ = Marshal.Release(ccw);

if (hr < 0)
{
    [MethodImpl(MethodImplOptions.NoInlining)]
    static Exception GetException(object obj, Guid iid, int hr)
    {
        // Special case 'E_NOINTERFACE' to provide a better exception message
        if (hr == ExceptionHelpers.E_NOINTERFACE)
        {
            return new InvalidCastException($"Failed to create a CCW for object of type '{obj.GetType()}' for interface with IID '{iid.ToString().ToUpperInvariant()}': the specified cast is not valid.");
        }

        return Marshal.GetExceptionForHR(hr);
    }

    throw GetException(obj, iid, hr);
}

return ppv;
```

In particular this also avoids the try/catch/rethrow, which is a bit faster, and also directly uses `Marshal.Release`.

**Review by Sergio0694 — COMMENTED** [review]
Did a first pass (haven't tested it yet), but looks good so far! 🙂

File: src/WinRT.Runtime/Attributes.cs [hunk 2]
```diff
@@ -266,6 +270,53 @@ public GeneratedBindableCustomPropertyAttribute(string[] propertyNames, Type[] i
...
+    /// is exposed to the WinRT ABI and needs a vtable generated for it.
+    /// The type which this attribute is placed on also needs to be marked partial.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Class, Inherited = false, AllowMultiple = false)]
+#if EMBED
+    internal
+#else
+    public
+#endif
+    sealed class GeneratedWinRTExposedTypeAttribute : Attribute
```
[the remark this answers is not in the document]
Comment (manodasanW) [reply src/WinRT.Runtime/Attributes.cs:287]: We can probably add one later but doing that also doesn't break anything as we do deduplication.  My concern would be we thinking it is a scenario where we would generate in, but we fail to.

File: docs/aot-trimming.md [hunk 3]
```diff
@@ -2,56 +2,58 @@
...
+## Source generator modes
+
+CsWinRT makes use of a source generator as part of making generated projections and uses of it trimming and AOT compatible. The source generator is primarily used to generate vtables for classes defined by the project and for generic types used such that if they were passed across the WinRT ABI, the interfaces that they implement will be projected into WinRT. CsWinRT still has its existing reflection based approach which still works for non AOT and non trimming scenarios but also comes at a runtime cost. There are four different modes which the source generator can run in which is determined by `CsWinRTAotOptimizerEnabled`:
+
+| Mode | Meaning |
+| ----- | ------- |
+| OptIn | Only generates code for attributed types. |
+| Auto | Attempts to detect scenarios that implement interfaces that can be passed across the WinRT ABI boundary and generates code for them. |
+| True* | If running in a project that references WinUI / UWP, then runs in `Auto`. Otherwise runs in `OptIn`. |
```
[the remark this answers is not in the document]
Comment (manodasanW) [reply docs/aot-trimming.md:17]: Right.

File: docs/aot-trimming.md [hunk 4]
```diff
@@ -2,56 +2,58 @@
...
+In `Auto` mode, it is recommended to set the `CsWinRTAotWarningLevel` to 2 and go through all the warnings and mark such types `partial` or suppress the warning if those types are not passed across the WinRT ABI. By default, if you reference the CsWinRT package, the `CsWinRTAotWarningLevel` is set to 1. If you don't have a CsWinRT package reference, then the .NET default is to set `CsWinRTAotWarningLevel` to 1 if you have marked your binary as AOT compatible either by using `IsAotCompatible` or `PublishAot`. Even though it is only by default enabled as warnings for AOT scenarios, you should manually set it to 1 or 2 if you support trimming and address them.
```
[the remark this answers is not in the document]
Comment (manodasanW) [reply docs/aot-trimming.md:36]: Yea, we can update that in the other PR.

**Pull request merged**